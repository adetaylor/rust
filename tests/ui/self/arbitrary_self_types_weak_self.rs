// run-pass

#![feature(receiver_trait)]

use std::rc::{Rc, Weak};

struct Foo(usize);

impl Foo {
  fn foo(self: Weak<Self>) -> i32 { 3 }
  fn bar(self: &Weak<Self>) -> i32 { 4 }
  fn baz(self: &mut Weak<Self>) -> i32 { 5 }

  fn as_ptr(self: &Weak<Self>) -> i32 { 6 }
}

fn main() {
  let foo = Rc::new(Foo(3));
  let mut foo_weak = Rc::downgrade(&foo);

  assert_eq!(foo_weak.as_ptr(), 6);
  //~^ ambiguous targets

  assert_eq!(foo_weak.bar(), 4);
  assert_eq!(foo_weak.baz(), 5);
  assert_eq!(foo_weak.foo(), 3);
}
