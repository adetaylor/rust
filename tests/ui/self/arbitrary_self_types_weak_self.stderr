warning: std::rc::Weak::<T, A>::as_ptr and Foo::as_ptr are ambiguous targets of this method call; explicitly choose one
  --> $DIR/arbitrary_self_types_weak_self.rs:21:23
   |
LL |   fn as_ptr(self: &Weak<Self>) -> i32 { 6 }
   |   ----------------------------------- because we'll call this method
...
LL |   assert_eq!(foo_weak.as_ptr(), 6);
   |                       ^^^^^^
   |
   = note: the method in crate alloc won't be called
   = note: `#[warn(shadowed_method_picked)]` on by default
help: call as a function not a method
   |
LL |   assert_eq!(foo_weak.Foo::as_ptr(FOO, ARGUMENTS)(), 6);
   |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: call as a function not a method
   |
LL |   assert_eq!(foo_weak.std::rc::Weak::<T, A>::as_ptr(FOO, ARGUMENTS)(), 6);
   |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

warning: 1 warning emitted

